text/0033-view-hierarchy.md

```diff
@@ -0,0 +1,116 @@
+
+## (De)obfuscation
+As the view hierarchy may contain obfuscated attribute values (see `type` attribute in the example above) we need to process the attachment server-side, similar to what we already do for minidumps, [or profiles](https://github.com/getsentry/sentry/blob/cf71af372677487d7d0a7fd8ac9dd092f9596cf4/src/sentry/profiles/task.py#L350-L360).
+
+# Options Considered
+- Use a new envelope type instead of attachements: Whilst doable it's consideribly more work without adding any benefits. Using attachements has several advantages
+    - Attachments are billable by default
+    - We can immediately start sending and using this data, no need for large Relay updates
+    - Less new code in the SDKs when compared to adding a new envelope item, so less hassle to maintain/test
+    - We don't have to touch outcomes/rate limiting/processing code at all
+    - We always can eventually migrate to a new envelope item type if we feel the need to, that option is still left open.
+- The view hierarchy could be part of the Event, but because of the size limit, it is not recommended
+
+
+# Drawbacks
+
+- Increases event payload size
+- We've decided to not add the content of textfields, editables or labels because of PII.
+    - We may evaluate an options to allow PII.
+
+
+## Initial Proposal
+This is how the envelope item header should look like:
+
+```json
+{"type":"view_hierarchy","length":...,"content_type":"application/json"}\n
+```
+
+And the envelope item payload will be a JSON array that contains a view hierarchy for each available screen/window. 
+This is how it should look like: 
+
+```json
+[
+    {
+        "type": "element_full_qualified_class_name", 
+        "identifier": "element_identifier",
+        "children": [...], //An array of ui elements.
+        "width": 100, 
+        "height": 100,
+        "depth": 100, //if applies
+        "x": 0,
+        "y": 1,
+        "z": 2, //if applies.
+        "visible": true|false,
+        "alpha": 1, //A float number from 0 to 1, where 0 means totally transparent, and 1 totally opaque.
+        "{extra_properties}": "{property value}" //adicional information by platform
+    }
+]
+```
+
+This is just an example, each platform will define the required properties.
+
+# Options Considered
+
+- We could send the same information as attachment, but this impact on attachment quota. And also we need to rely on file name to determined whether an attachment is a view hierarchy or not, this is a problem we already have with [screenshots](https://develop.sentry.dev/sdk/features/#screenshots).
+- The view hierarchy could be part of the Event, but because of the size limit, isn't recommended. 
+
+# Drawbacks
+
+- This can significantly increase the envelope size. 
+- Obfuscation is a problem for custom UI controls, we'll probably need ingestion to de-obfuscate in the server side.
```

Comment:
Link this to https://docs.sentry.io/product/data-management-settings/scrubbing/attachment-scrubbing/